src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -231,21 +237,270 @@ def _estimate_inplane_rotations(
+
+        Given the common lines matrix `clmatrix`, a list of images specified in k_list
+        and the number of common lines n_theta, find the (i, j) rotation block Rij.
+        :param clmatrix: The common lines matrix
+        :param i: The i image
+        :param j: The j image
+        :param k_list: The list of images for the third image for voting algorithm
+        :param n_theta: The number of points in the theta direction (common lines)
+        :return: The (i,j) rotation block of the synchronization matrix
+        """
+        good_k = self._vote_ij(clmatrix, n_theta, i, j, k_list)
+
+        rots = self._rotratio_eulerangle_vec(clmatrix, i, j, good_k, n_theta)
+
+        if rots is not None:
+            rot_mean = np.mean(rots, 0)
+
+        else:
+            # This is for the case that images i and j correspond to the same
+            # viewing direction and differ only by in-plane rotation.
+            # We set to zero as in the Matlab code.
+            rot_mean = np.zeros((3, 3))
+
+        return rot_mean
+
+    def _local_J_sync_c3_c4(self, Rijs, Riis):
+        """
+        Estimate viis and vijs. In order to estimate vij = vi @ vj.T, it is necessary for Rii, Rjj,
+        and Rij to be of the same handedness. We perform a local handedness synchronization and
+        set vij = 1/n*sum(Rii^s @ Rij @ Rjj^s).
+
+        :param Rijs: An n-choose-2x3x3 array of estimates of relative rotations
+            (each pair of images induces two estimates).
+        :param Riis: A nx3x3 array of estimates of self-relative rotations.
+        :return: vijs, viis
+        """
+        order = self.order
+        n_img = self.n_img
+
+        nchoose2 = int(n_img * (n_img - 1) / 2)
+        assert (
+            len(Riis) == n_img
+        ), f"There must be one self-relative rotation per image. Got {len(Riis)} Riis."
+        assert (
+            len(Rijs) == nchoose2
+        ), f"There must be n-choose-2 relative rotations. Got {len(Rijs)}."
+
+        # Estimate viis from Riis. vii = 1/order * sum(Rii^s) for s = 0, 1, ..., order.
+        viis = np.zeros((n_img, 3, 3))
+        for i, Rii in enumerate(Riis):
+            viis[i] = np.mean(
+                [np.linalg.matrix_power(Rii, s) for s in np.arange(order)], axis=0
+            )
+
+        # Estimate vijs via local handedness synchronization.
+        vijs = np.zeros((nchoose2, 3, 3))
+        e1 = [1, 0, 0]
+        opts = np.zeros((8, 3, 3))
+        scores_rank1 = np.zeros(8)
+        min_idxs = np.zeros((nchoose2, 3, 3))
+        pairs = all_pairs(n_img)
+        for idx, (i, j) in enumerate(pairs):
+            Rii = Riis[i]
+            Rjj = Riis[j]
+            Rij = Rijs[idx]
+
+            Rii_J = J_conjugate(Rii)
+            Rjj_J = J_conjugate(Rjj)
+
+            # vij should be a singular matrix.
+            # We test 8 combinations of handedness and rotation by {g, g^n-1} for singularity to determine:
+            # a. whether to transpose Rii
```

Comment:
So which corresponds to which? a → 4, 5, 6, 7, b → 1, 3, 5, 7, c → 2, 3, 6, 7?

There's also a way to loop this by defining `opts = np.zeros(2, 2, 2, 3, 3)`, then doing
```
for ii_trans in [0, 1]:
    for ii_conj in [0, 1]:
        for jj_conj in [0, 1]:
             _Rii = Rii.T if ii_trans else Rii
             _Rii = J_conjugate(_Rii) if ii_conj else _Rii
             _Rjj = J_conjugate(Rjj) if jj_conj else Rjj
             opts[ii_trans, ii_conj, jj_conj] = (Rij + _Rii @ Rij @ _Rjj + _Rii @ Rij @ _Rjj) / 3
```

Reply:
Looks great. Thanks!